File: AllSamples/Samples/Samples_Source/media services/consoleplayer/consoleplayercs/consoleplayer/consoleplayer.cs
//---------------------------------------------------------------------
//
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------

using System;
using System.Windows.Media.Core;
using System.Windows.Media.Types;

namespace ConsolePlayerC
{
    /// <summary>
    /// This simple console application plays the audio file specified on the command line.
    /// </summary>
    class MediaEngineDemo
    {
        private MediaEngine m_mediaEngine = null;
        private System.Threading.AutoResetEvent engineCanBeDisposedEvent = null;
        private System.Threading.AutoResetEvent engineOpenForPlayEvent = null;
        private Boolean m_validFile = true;

        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Main(string[] args)
        {
            if (0 < args.Length)
            {
                MediaEngineDemo player = new MediaEngineDemo(args[0]);
            }
            else
            {
                System.Console.WriteLine("No media file specified.");
            }
        }

        MediaEngineDemo(string mediaFilePath)
        {
            // Create MediaEngine.
            m_mediaEngine = new MediaEngine();

            // Add event handlers.
            m_mediaEngine.MediaEnded += new System.Windows.Media.Core.MediaEventHandler(this.MediaEnded);
            m_mediaEngine.MediaOpened += new System.Windows.Media.Core.MediaEventHandler(this.MediaOpened);

            // Create private events to be signaled by media event handlers.
            engineCanBeDisposedEvent = new System.Threading.AutoResetEvent(false);
            engineOpenForPlayEvent = new System.Threading.AutoResetEvent(false);

            // Open and play the file. 
                // Because media engine methods are asynchronous, they always return "success" 
                // and any errors must be caught in the event handlers. In this case, if the file
                // is not valid, a global variable is set.
                m_mediaEngine.Open(mediaFilePath, null, typeof(AudioMediaType));
                engineOpenForPlayEvent.WaitOne();
            if (m_validFile)
            {
                m_mediaEngine.Start();
                // Wait till playback is finished before exiting.
                engineCanBeDisposedEvent.WaitOne();
                }
            else
            {
                Console.WriteLine("Could not play file.");
            }
            engineCanBeDisposedEvent.Close();
            engineOpenForPlayEvent.Close();
        }

        /// <summary>
        /// Handler for MediaEnded event. This event is signaled when playback has reached
        /// the end of the file.
        /// </summary>
        private void MediaEnded(Object sender, MediaEventArgs args)
        {
            engineCanBeDisposedEvent.Set();
        }

        /// <summary>
        /// Handler for MediaOpened event. Any exception raised by trying to open an invalid file
        /// must be caught here.
        /// </summary>
        private void MediaOpened(Object sender, MediaEventArgs args)
        {
            if (args.Exception != null)
            {
                m_validFile = false;
            }
            engineOpenForPlayEvent.Set();
        }
        
    }
}

File: AllSamples/Samples/Samples_Source/media services/consoleplayer/consoleplayervb/consoleplayer/consoleplayer.vb
' '---------------------------------------------------------------------
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
' This source code is intended only as a supplement to Microsoft
' Development Tools and/or on-line documentation.  See these other
' materials for detailed information regarding Microsoft code samples.
' 
' THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
' KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
' IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
' PARTICULAR PURPOSE.
'---------------------------------------------------------------------

Imports System
Imports System.Windows.Media.Core
Imports System.Windows.Media.Types


' This console application plays the sound file specified on the command line.

Class MediaEngineDemo

    Private m_mediaEngine As MediaEngine
    Private engineCanBeDisposedEvent As System.Threading.AutoResetEvent
    Private engineOpenCompleteEvent As System.Threading.AutoResetEvent
    Private m_validFile As Boolean = True

    Shared Sub Main(ByVal args() As String)
        If (0 < args.Length) Then
            Dim player As MediaEngineDemo
            player = New MediaEngineDemo(args(0))
        Else
            Console.WriteLine("No media file specified.")
        End If
    End Sub

    Private Sub New(ByVal mediaFilePath As String)
        ' Create MediaEngine.
        m_mediaEngine = New MediaEngine

        ' Add event handlers.
        AddHandler m_mediaEngine.MediaEnded, AddressOf MediaEnded
        AddHandler m_mediaEngine.MediaOpened, AddressOf MediaOpened

        ' Create private events to be signaled by media event handlers.
        engineCanBeDisposedEvent = New System.Threading.AutoResetEvent(False)
        engineOpenCompleteEvent = New System.Threading.AutoResetEvent(False)

        ' Open and play the file. 
        ' Because media engine methods are asynchronous, they always return "success" 
        ' and any errors must be caught in the event handlers. In this case, if the file
        ' is not valid, a global variable is set.
        m_mediaEngine.Open(mediaFilePath, Nothing, GetType(AudioMediaType))
        engineOpenCompleteEvent.WaitOne()
        If (m_validFile = True) Then

            ' No exception, so start playing.
            m_mediaEngine.Start()

            ' Wait till playback is finished before exiting.
            engineCanBeDisposedEvent.WaitOne()
        Else
            Console.WriteLine("Could not open file.")
        End If
        engineCanBeDisposedEvent.Close()
        engineOpenCompleteEvent.Close()
    End Sub

    Private Sub MediaEnded(ByVal sender As Object, ByVal args As MediaEventArgs)
        ' Set a private event to notify the main thread that it is okay to exit.
        engineCanBeDisposedEvent.Set()
    End Sub

    Private Sub MediaOpened(ByVal sender As Object, ByVal args As MediaEventArgs)
        If (Not args.Exception Is Nothing) Then
            m_validFile = False
        End If
        engineOpenCompleteEvent.Set()
    End Sub

End Class


